File: src/lib/ai/coachWatch.ts
/**
 * Deterministischer Assistenten-Monitor.
 * Kein LLM: Engines entscheiden, der Chat formuliert höchstens um.
 */

import { analyzePostRide } from "@/lib/ai/postRideAnalysis";
import { normalizeBikes, normalizeRides } from "@/lib/ai/normalize";
import { checkBikeCompatibility } from "@/lib/compatibility/engine";
import { buildMaintenanceAlerts } from "@/lib/home/maintenanceAlerts";
import type { RangeCalibration } from "@/lib/ebike/range";
import type {
  Bike,
  MaintenanceInterval,
  Ride,
  RideFeedback,
  RiderProfile,
} from "@/types";
import type { ChromeLang } from "@/lib/i18n/chromeLang";
import { chatCopy, type ChatCopy } from "@/lib/i18n/chatCopy";
import {
  postRideAnalysisCopy,
  type PostRideAnalysisCopy,
} from "@/lib/i18n/postRideAnalysisCopy";

export type CoachKind =
  | "maintenance"
  | "wear"
  | "compat"
  | "setup"
  | "range"
  | "feedback";

export type CoachSeverity = "info" | "due_soon" | "overdue";

export type CoachToolName =
  | "watch"
  | "garage"
  | "compat"
  | "setup_history"
  | "ride_stats"
  | "route_search"
  | "product_search"
  | "range";

export interface CoachNumber {
  value: number;
  unit: string;
  source: string;
}

export interface CoachNotice {
  id: string;
  kind: CoachKind;
  severity: CoachSeverity;
  title: string;
  detail: string;
  reasoning: string;
  href: string;
  bikeId?: string;
  tool: CoachToolName;
  query: string;
  fingerprint: string;
  numbers: CoachNumber[];
}

export interface CoachWatchInput {
  bikes: Bike[];
  rides: Ride[];
  intervals?: MaintenanceInterval[];
  profile: RiderProfile;
  calibration?: RangeCalibration | null;
  rideFeedbacks?: RideFeedback[];
  now?: Date;
}

const FEEDBACK_WINDOW_MS = 48 * 60 * 60 * 1000;
const MAX_NOTICES = 8;

function fingerprint(parts: string[]): string {
  return parts.join("|").slice(0, 180);
}

function rank(s: CoachSeverity): number {
  if (s === "overdue") return 0;
  if (s === "due_soon") return 1;
  return 2;
}

export function buildCoachWatch(input: CoachWatchInput): CoachNotice[] {
  const bikes = normalizeBikes(input.bikes);
  const rides = normalizeRides(input.rides);
  const intervals = input.intervals ?? [];
  const now = input.now ?? new Date();
  if (bikes.length === 0) return [];

  const notices: CoachNotice[] = [];

  for (const bike of bikes) {
    try {
      const alerts = buildMaintenanceAlerts({
        bike,
        rides: rides.filter((r) => !r.bikeId || r.bikeId === bike.id),
        intervals: intervals.filter((i) => i.bikeId === bike.id),
        max: 4,
      });
      for (const a of alerts) {
        const kind: CoachKind = a.id.startsWith("wear-") ? "wear" : "maintenance";
        notices.push({
          id: `maint:${bike.id}:${a.id}`,
          kind,
          severity: a.severity,
          title: a.title,
          detail: `${bike.name}: ${a.detail}`,
          reasoning: a.reasoning,
          href: "/garage?tab=maintenance",
          bikeId: bike.id,
          tool: kind === "wear" ? "product_search" : "garage",
          query: a.title,
          fingerprint: fingerprint([a.severity, a.title, a.detail]),
          numbers: [],
        });
      }
    } catch {
      /* unvollständige Garage */
    }

    try {
      const results = checkBikeCompatibility(bike).filter(
        (r) => r.verdict === "INCOMPATIBLE"
      );
      for (const r of results.slice(0, 2)) {
        const overdue = r.severity === "safety_critical";
        notices.push({
          id: `compat:${bike.id}:${r.ruleCode}`,
          kind: "compat",
          severity: overdue ? "overdue" : "due_soon",
          title: r.title,
          detail: r.explainDe,
          reasoning: `${r.ruleCode} · ${bike.name}`,
          href: "/garage",
          bikeId: bike.id,
          tool: "compat",
          query: `Passt ${r.title}?`,
          fingerprint: fingerprint([r.verdict, r.ruleCode, r.explainDe]),
          numbers: [],
        });
      }
    } catch {
      /* */
    }

    if (bike.isEbike) {
      const samples = input.calibration?.samples ?? 0;
      if (samples < 2) {
        notices.push({
          id: `range:${bike.id}:cal`,
          kind: "range",
          severity: "info",
          title: "Reichweite noch unsicher",
          detail: `${bike.name}: nach ein paar Fahrten mit SOC wird die Spanne enger.`,
          reasoning: `Kalibrierung ${samples} Sample(s) — Physikmodell ohne Selbstabgleich.`,
          href: "/chat",
          bikeId: bike.id,
          tool: "range",
          query: "Welche Reichweite habe ich mit aktuellem Akku?",
          fingerprint: fingerprint(["range", String(samples)]),
          numbers: [{ value: samples, unit: "", source: "range.samples" }],
        });
      }
    }
  }

  const last = [...rides].sort(
    (a, b) => new Date(b.startTime).getTime() - new Date(a.startTime).getTime()
  )[0];
  if (last) {
    const start = new Date(last.startTime).getTime();
    const age = now.getTime() - start;
    if (age >= 0 && age <= FEEDBACK_WINDOW_MS) {
      const fb = (input.rideFeedbacks ?? []).find((f) => f.rideId === last.id);
      if (!fb || fb.skipped) {
        notices.push({
          id: `feedback:${last.id}`,
          kind: "feedback",
          severity: "info",
          title: "Kurzes Feedback zur letzten Fahrt",
          detail: "Drei Taps — der Assistent nutzt das für Setup-Hinweise.",
          reasoning: "Post-Ride-Fenster 48 h, noch kein Feedback.",
          href: "/post-ride",
          bikeId: last.bikeId,
          tool: "setup_history",
          query: "Was hat sich am Setup nach der letzten Fahrt angeboten?",
          fingerprint: fingerprint(["feedback", last.id]),
          numbers: [{ value: 48, unit: "h", source: "feedback.window" }],
        });
      }

      const bike = last.bikeId
        ? bikes.find((b) => b.id === last.bikeId)
        : bikes[0];
      if (bike) {
        try {
          const current = bike.setups.find((s) => s.isCurrent);
          const analysis = analyzePostRide({
            ride: last,
            bike,
            setup: current,
            feedback: fb && !fb.skipped ? fb : undefined,
          });
          const sug = analysis.setupSuggestion;
          if (sug) {
            notices.push({
              id: `setup:${last.id}:${sug.adjusterKey ?? sug.title}`,
              kind: "setup",
              severity: sug.confidence === "high" ? "due_soon" : "info",
              title: sug.title,
              detail: sug.content,
              reasoning: sug.reasoning,
              href: "/post-ride",
              bikeId: bike.id,
              tool: "setup_history",
              query: sug.title,
              fingerprint: fingerprint([sug.title, sug.content]),
              numbers:
                sug.suggestedDelta != null
                  ? [
                      {
                        value: sug.suggestedDelta,
                        unit: "klicks",
                        source: "setup.delta",
                      },
                    ]
                  : [],
            });
          }
        } catch {
          /* */
        }
      }
    }
  }

  notices.sort((a, b) => rank(a.severity) - rank(b.severity));
  const seen = new Set<string>();
  const unique: CoachNotice[] = [];
  for (const n of notices) {
    if (seen.has(n.id)) continue;
    seen.add(n.id);
    unique.push(n);
    if (unique.length >= MAX_NOTICES) break;
  }
  return unique;
}

export function localizeCoachNotice(
  n: CoachNotice,
  lang: ChromeLang = "de"
): { title: string; detail: string } {
  const chat = chatCopy(lang);
  const analysis = postRideAnalysisCopy(lang);
  return localizeCoachNoticeWith(n, chat, analysis);
}

function localizeCoachNoticeWith(
  n: CoachNotice,
  chat: ChatCopy,
  analysis: PostRideAnalysisCopy
): { title: string; detail: string } {
  if (n.kind === "feedback") {
    return { title: chat.feedbackTitle, detail: chat.feedbackDetail };
  }
  if (n.kind === "setup") {
    if (n.title.includes("2 Klicks langsamer")) {
      return { title: analysis.sugReboundSlowTitle, detail: n.detail };
    }
    if (n.title.includes("2 Klicks schneller")) {
      return { title: analysis.sugReboundFastTitle, detail: n.detail };
    }
    if (/Luftdruck|air pressure|pression|pressione|luchtdruk/i.test(n.title)) {
      return { title: analysis.sugPressureTitle, detail: n.detail };
    }
  }
  return { title: n.title, detail: n.detail };
}

export function formulateCoachWatch(
  notices: CoachNotice[],
  lang: ChromeLang = "de"
): string {
  const chat = chatCopy(lang);
  const analysis = postRideAnalysisCopy(lang);
  if (notices.length === 0) return chat.watchNothing;
  const overdue = notices.filter((n) => n.severity === "overdue").length;
  const soon = notices.filter((n) => n.severity === "due_soon").length;
  const lines = notices.slice(0, 5).map((n) => {
    const loc = localizeCoachNoticeWith(n, chat, analysis);
    const tag =
      n.severity === "overdue"
        ? chat.sevOverdue
        : n.severity === "due_soon"
          ? chat.sevSoon
          : chat.sevInfo;
    return `${tag}: ${loc.title} — ${loc.detail}`;
  });
  const head =
    overdue > 0
      ? `${overdue} ${chat.sevOverdue.toLowerCase()}, ${soon} ${chat.sevSoon.toLowerCase()}.`
      : soon > 0
        ? `${soon} ${chat.sevSoon.toLowerCase()}.`
        : `${notices.length} ${chat.sevInfo}.`;
  return `${head} ${lines.join(" ")}`;
}

File: mobile/lib/domain/ai/coach_watch.dart
// Deterministischer Assistenten-Monitor — Spiegel src/lib/ai/coachWatch.ts.

import '../bike.dart';
import '../compatibility/engine.dart';
import '../compatibility/rules.dart';
import '../component.dart';
import '../ebike/range.dart';
import '../maintenance/intervals.dart';
import '../post_ride/analyze.dart';
import '../ride.dart';
import '../setup.dart';

enum CoachKind { maintenance, wear, compat, setup, range, feedback }

enum CoachSeverity { info, dueSoon, overdue }

class CoachNotice {
  const CoachNotice({
    required this.id,
    required this.kind,
    required this.severity,
    required this.title,
    required this.detail,
    required this.reasoning,
    required this.href,
    required this.tool,
    required this.query,
    required this.fingerprint,
    this.bikeId,
    this.numbers = const [],
  });

  final String id;
  final CoachKind kind;
  final CoachSeverity severity;
  final String title;
  final String detail;
  final String reasoning;
  final String href;
  final String? bikeId;
  final String tool;
  final String query;
  final String fingerprint;
  final List<Map<String, dynamic>> numbers;

  Map<String, dynamic> toJson() => {
        'id': id,
        'kind': kind.name,
        'severity': switch (severity) {
          CoachSeverity.overdue => 'overdue',
          CoachSeverity.dueSoon => 'due_soon',
          CoachSeverity.info => 'info',
        },
        'title': title,
        'detail': detail,
        'reasoning': reasoning,
        'href': href,
        if (bikeId != null) 'bikeId': bikeId,
        'tool': tool,
        'query': query,
        'fingerprint': fingerprint,
        'numbers': numbers,
      };
}

class CoachWatchInput {
  const CoachWatchInput({
    required this.bikes,
    required this.componentsByBike,
    required this.rides,
    this.setupsByBike = const {},
    this.calibration,
    this.now,
  });

  final List<Bike> bikes;
  final Map<String, List<BikeComponent>> componentsByBike;
  final List<RideRecord> rides;
  final Map<String, List<BikeSetup>> setupsByBike;
  final RangeCalibration? calibration;
  final DateTime? now;
}

const _feedbackWindow = Duration(hours: 48);
const _maxNotices = 8;

int _rank(CoachSeverity s) => switch (s) {
      CoachSeverity.overdue => 0,
      CoachSeverity.dueSoon => 1,
      CoachSeverity.info => 2,
    };

List<CoachNotice> buildCoachWatch(CoachWatchInput input) {
  if (input.bikes.isEmpty) return [];
  final now = input.now ?? DateTime.now().toUtc();
  final notices = <CoachNotice>[];

  for (final bike in input.bikes) {
    final comps = input.componentsByBike[bike.id] ?? const <BikeComponent>[];
    final due = listDueMaintenance(bike: bike, components: comps, now: now);
    for (final a in due) {
      notices.add(
        CoachNotice(
          id: 'maint:${bike.id}:${a.slot.apiId}:${a.label}',
          kind: CoachKind.maintenance,
          severity: a.status == DueStatus.overdue
              ? CoachSeverity.overdue
              : CoachSeverity.dueSoon,
          title: a.label,
          detail: '${bike.name}: ${a.remainingLabel}',
          reasoning: a.sourceLabel ?? 'Intervall',
          href: '/garage',
          bikeId: bike.id,
          tool: 'garage',
          query: a.label,
          fingerprint: '${a.status.name}|${a.label}|${a.remainingLabel}',
        ),
      );
    }

    try {
      final results = checkBikeCompatibility(comps)
          .where((r) => r.verdict == CompatVerdict.incompatible)
          .take(2);
      for (final r in results) {
        final overdue = r.severity == RuleSeverity.safetyCritical;
        notices.add(
          CoachNotice(
            id: 'compat:${bike.id}:${r.ruleCode}',
            kind: CoachKind.compat,
            severity:
                overdue ? CoachSeverity.overdue : CoachSeverity.dueSoon,
            title: r.title,
            detail: r.explainDe,
            reasoning: '${r.ruleCode} · ${bike.name}',
            href: '/garage',
            bikeId: bike.id,
            tool: 'compat',
            query: 'Passt ${r.title}?',
            fingerprint: '${r.verdict.name}|${r.ruleCode}|${r.explainDe}',
          ),
        );
      }
    } catch (_) {}

    if (bike.hasElectricAssist) {
      final samples = input.calibration?.samples ?? 0;
      if (samples < 2) {
        notices.add(
          CoachNotice(
            id: 'range:${bike.id}:cal',
            kind: CoachKind.range,
            severity: CoachSeverity.info,
            title: 'Reichweite noch unsicher',
            detail:
                '${bike.name}: nach ein paar Fahrten mit SOC wird die Spanne enger.',
            reasoning: 'Kalibrierung $samples Sample(s).',
            href: '/chat',
            bikeId: bike.id,
            tool: 'range',
            query: 'Welche Reichweite habe ich mit aktuellem Akku?',
            fingerprint: 'range|$samples',
            numbers: [
              {'value': samples, 'unit': '', 'source': 'range.samples'},
            ],
          ),
        );
      }
    }
  }

  final sortedRides = [...input.rides]
    ..sort((a, b) => b.startedAt.compareTo(a.startedAt));
  if (sortedRides.isNotEmpty) {
    final last = sortedRides.first;
    final age = now.difference(last.endedAt ?? last.startedAt);
    if (!age.isNegative && age <= _feedbackWindow) {
      final fb = last.feedback;
      if (fb == null || fb.skipped) {
        notices.add(
          CoachNotice(
            id: 'feedback:${last.id}',
            kind: CoachKind.feedback,
            severity: CoachSeverity.info,
            title: 'Kurzes Feedback zur letzten Fahrt',
            detail:
                'Drei Taps — der Assistent nutzt das für Setup-Hinweise.',
            reasoning: 'Post-Ride-Fenster 48 h, noch kein Feedback.',
            href: '/post-ride',
            bikeId: last.bikeId,
            tool: 'setup_history',
            query:
                'Was hat sich am Setup nach der letzten Fahrt angeboten?',
            fingerprint: 'feedback|${last.id}',
            numbers: [
              {'value': 48, 'unit': 'h', 'source': 'feedback.window'},
            ],
          ),
        );
      }
      Bike? bike;
      for (final b in input.bikes) {
        if (b.id == last.bikeId) {
          bike = b;
          break;
        }
      }
      bike ??= input.bikes.isEmpty ? null : input.bikes.first;
      if (bike != null) {
        final setups = input.setupsByBike[bike.id] ?? const <BikeSetup>[];
        BikeSetup? current;
        for (final s in setups) {
          if (s.isCurrent) current = s;
        }
        final rebound = current?.valueFor('fork.rebound')?.round();
        final analysis = analyzePostRide(
          ride: last,
          bikeName: bike.name,
          feedback: fb,
          forkReboundClicks: rebound,
        );
        final sug = analysis.setupSuggestion;
        if (sug != null) {
          notices.add(
            CoachNotice(
              id: 'setup:${last.id}:${sug.adjusterKey ?? sug.title}',
              kind: CoachKind.setup,
              severity: sug.confidence == 'high'
                  ? CoachSeverity.dueSoon
                  : CoachSeverity.info,
              title: sug.title,
              detail: sug.content,
              reasoning: sug.reasoning,
              href: '/post-ride',
              bikeId: bike.id,
              tool: 'setup_history',
              query: sug.title,
              fingerprint: '${sug.title}|${sug.content}',
              numbers: [
                if (sug.suggestedDelta != null)
                  {
                    'value': sug.suggestedDelta,
                    'unit': 'klicks',
                    'source': 'setup.delta',
                  },
              ],
            ),
          );
        }
      }
    }
  }

  notices.sort((a, b) => _rank(a.severity).compareTo(_rank(b.severity)));
  final seen = <String>{};
  final unique = <CoachNotice>[];
  for (final n in notices) {
    if (!seen.add(n.id)) continue;
    unique.add(n);
    if (unique.length >= _maxNotices) break;
  }
  return unique;
}

String formulateCoachWatch(List<CoachNotice> notices) {
  if (notices.isEmpty) {
    return 'Gerade steht nichts an. Wartung, Verschleiß, Setup und '
        'Kompatibilität sind im grünen Bereich.';
  }
  final overdue =
      notices.where((n) => n.severity == CoachSeverity.overdue).length;
  final soon =
      notices.where((n) => n.severity == CoachSeverity.dueSoon).length;
  final lines = notices.take(5).map((n) {
    final tag = switch (n.severity) {
      CoachSeverity.overdue => 'überfällig',
      CoachSeverity.dueSoon => 'bald',
      CoachSeverity.info => 'Hinweis',
    };
    return '$tag: ${n.title} — ${n.detail}';
  });
  final head = overdue > 0
      ? '$overdue überfällig, $soon bald fällig.'
      : soon > 0
          ? '$soon bald fällig.'
          : '${notices.length} Hinweis${notices.length == 1 ? '' : 'e'}.';
  return '$head ${lines.join(' ')}';
}
